### ISSUE TITLE
Universal virtualfile_from_data function to replace virtualfile_from_grid/matrix/vectors

### ISSUE THREAD

**Issue by weiji14** [GenericMappingTools/pygmt#949]
**Description of the desired feature**

As a user, I'd like to be able to pass in any type of data into a PyGMT function in a consistent way, be it:

- a `str` type filename
- a `numpy.ndarray`
- a `pandas.DataFrame` matrix
- a `geopandas.GeoDataFrame` (#608)
- an `xarray.DataArray` grid, or even 1d arrays!
- etc

### Current state

We have a not so nice `if... then` block that currently handles the Input/Output (I/O) of data into PyGMT. This results in an incosistent API. E.g. [`fig.plot`](https://www.pygmt.org/v0.3.0/api/generated/pygmt.Figure.plot.html) requires `numpy.ndarray` inputs while [`pygmt.grdtrack`](https://www.pygmt.org/v0.3.0/api/generated/pygmt.grdtrack.html) requires `pandas.DataFrame` inputs. Users (e.g. at https://forum.generic-mapping-tools.org/t/pygmt-plot-errorbars) should not have to figure out which [PyData](https://pydata.org/) format is needed.

_Originally posted by @weiji14 in https://github.com/GenericMappingTools/pygmt/issues/946#issuecomment-782925067_

> I was going to create a `virtualfile_from_data()` function to replace this common block of code in our functions:

> ```python
>     with Session() as lib:
>         # Choose how data will be passed in to the module
>         if kind == "file":
>             file_context = dummy_context(data)
>         elif kind == "matrix":
>             file_context = lib.virtualfile_from_matrix(data)
>         elif kind == "vectors":
>             file_context = lib.virtualfile_from_vectors(...)
> ```

### Future state

The idea is to have a `virtualfile_from_data()` function (open to other names) that wraps around the current `dummy_context`/`virtualfile_from_matrix`/`virtualfile_from_vectors`/`virtualfile_from_grid` functions, which makes things easier for both end users and developers wrapping new modules that handle data.

This universal `virtualfile_from_data()` may pave the way to support more data inputs in the future:

- `geopandas.GeoDataFrame` inputs, useful for `fig.plot` #608 
- ObsPy [`Trace`](https://docs.obspy.org/packages/autogen/obspy.core.trace.Trace.html)/[`Stream`](https://docs.obspy.org/packages/autogen/obspy.core.stream.Stream.html)/ [`Inventory`](https://docs.obspy.org/packages/autogen/obspy.core.inventory.inventory.Inventory.html) objects #967
- `io.StringIO`, useful for `fig.legend` #576
- [`pint`](https://github.com/hgrecco/pint) arrays? #735
- other OGR/GDAL formats?
- etc

**TLDR**: Be able to throw any sort of data into a PyGMT function and it will 'just work'.

Note that `xarray` is in the process of refactoring to an APIv2 (see https://github.com/pydata/xarray/issues/4309) that has some similarities with this issue. They call it a 'Flexible Backend", a sort of extensible 'plugin'-type system which is a bit more future proofed to allow for new data formats (e.g. GPU backed arrays) coming in.

**Are you willing to help implement and maintain this feature?** Yes, but help is very welcome

**Update**: The `virtualfile_from_data` function has been created in #961. Feel free people to refactor some of the plotting/data processing functions in https://github.com/GenericMappingTools/pygmt/tree/master/pygmt/src to use this (e.g. [`contour`](https://github.com/GenericMappingTools/pygmt/blob/v0.3.0/pygmt/src/contour.py), [`meca`](https://github.com/GenericMappingTools/pygmt/blob/v0.3.0/pygmt/src/meca.py), etc).

TODO:
- [x] contour (#1531)
- [x] grdtrack (#1189)
- [x] meca (https://github.com/GenericMappingTools/pygmt/pull/1784)
- [x] surface
- [x] text (#1121)
- [ ] grdview/grdimage (#1988)

**Comment by liamtoney** [comment]
I'd add `Path` objects if they're not already supported. :)
